go/adbc/driver/bigquery/record_reader.go

```diff
@@ -0,0 +1,319 @@
+	}
+	return ctx.Err()
+}
+
+func runQuery(ctx context.Context, query *bigquery.Query, executeUpdate bool) (bigquery.ArrowIterator, int64, error) {
+	job, err := query.Run(ctx)
+	if err != nil {
+		return nil, -1, err
+	}
+	if executeUpdate {
+		return nil, 0, nil
+	} else {
+		iter, err := job.Read(ctx)
+		if err != nil {
+			return nil, -1, err
+		}
+		arrowIterator, err := iter.ArrowIterator()
+		if err != nil {
+			return nil, -1, err
+		}
+		totalRows := int64(iter.TotalRows)
+		return arrowIterator, totalRows, nil
+	}
+}
+
+func ipcReaderFromArrowIterator(arrowIterator bigquery.ArrowIterator, alloc memory.Allocator) (*ipc.Reader, error) {
+	arrowItReader := bigquery.NewArrowIteratorReader(arrowIterator)
+	rdr, err := ipc.NewReader(arrowItReader, ipc.WithAllocator(alloc))
+	if err != nil {
+		return nil, err
+	}
+	return rdr, nil
+}
+
+func getQueryParameter(values arrow.Record, row int, parameterMode string) ([]bigquery.QueryParameter, error) {
+	parameters := make([]bigquery.QueryParameter, values.NumCols())
+	includeName := parameterMode == OptionValueQueryParameterModeNamed
+	schema := values.Schema()
+	for i, v := range values.Columns() {
+		pi, err := arrowValueToQueryParameterValue(schema.Field(i), v, row)
+		if err != nil {
+			return nil, err
+		}
+		parameters[i] = pi
+		if includeName {
+			parameters[i].Name = values.ColumnName(i)
+		}
+	}
+	return parameters, nil
+}
+
+func runPlainQuery(ctx context.Context, query *bigquery.Query, alloc memory.Allocator, resultRecordBufferSize int) (bigqueryRdr *reader, totalRows int64, err error) {
+	arrowIterator, totalRows, err := runQuery(ctx, query, false)
+	if err != nil {
+		return nil, -1, err
+	}
+	rdr, err := ipcReaderFromArrowIterator(arrowIterator, alloc)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	chs := make([]chan arrow.Record, 1)
+	ctx, cancelFn := context.WithCancel(ctx)
+	ch := make(chan arrow.Record, resultRecordBufferSize)
+	chs[0] = ch
+
+	defer func() {
+		if err != nil {
+			close(ch)
+			cancelFn()
+		}
+	}()
+
+	bigqueryRdr = &reader{
+		refCount:   1,
+		chs:        chs,
+		curChIndex: 0,
+		err:        nil,
+		cancelFn:   cancelFn,
+		schema:     rdr.Schema(),
+	}
+
+	go func() {
+		defer rdr.Release()
+		for rdr.Next() && ctx.Err() == nil {
+			rec := rdr.Record()
+			rec.Retain()
+			ch <- rec
+		}
+
+		err = checkContext(ctx, rdr.Err())
+		defer close(ch)
+	}()
+
+	return bigqueryRdr, totalRows, nil
+}
+
+// kicks off a goroutine for each endpoint and returns a reader which
+// gathers all of the records as they come in.
+func newRecordReader(ctx context.Context, query *bigquery.Query, boundParameters array.RecordReader, parameterMode string, alloc memory.Allocator, resultRecordBufferSize, prefetchConcurrency int) (bigqueryRdr *reader, totalRows int64, err error) {
+	if boundParameters == nil {
+		return runPlainQuery(ctx, query, alloc, resultRecordBufferSize)
+	} else {
+		recs := make([]arrow.Record, 0)
```

Comment:
same comment as before, drop the else and unindent everything since we just return above it. it'll make the code easier to read